official/recommendation/ncf.py

```diff
@@ -0,0 +1,235 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""NCF framework to train and evaluate the NeuMF model.
+
+The NeuMF model ensembles both MF and MLP models under the NCF framework. Check
+`neumf_model.py` for more details about the models.
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import ast
+import heapq
+import math
+import os
+import sys
+import time
+
+import tensorflow as tf  # pylint: disable=g-bad-import-order
+
+import data_download
+import ncf_dataset
+import neumf_model
+import numpy as np
+from six.moves import xrange  # pylint: disable=redefined-builtin
+
+
+def evaluate_model(est_model, user_input, item_input, gt_items, top_k):
+  """Model evaluation with HR and NDCG metrics.
+
+  The evaluation protocol is to rank the test interacted item among the randomly
+  chosen 100 items that are not interacted by the user. The performance of the
+  ranked list is judged by Hit Ratio (HR) and Normalized Discounted Cumulative
+  Gain (NDCG).
+
+  For evaluation, the ranked list is truncated at 10 for both metrics. As such,
+  the HR intuitively measures whether the test item is present on the top-10
+  list, and the NDCG accounts for the position of the hit by assigning higher
+  scores to hits at top ranks. Both metrics are calculated for each test user,
+  and the average scores are reported.
+
+  Args:
+    est_model: The Estimator.
+    user_input: The user input for evaluation.
+    item_input: The item input for evaluation.
+    gt_items: The test item for HR and NDCG calculation.
+    top_k: The top-k list. It is 10 by default.
+
+  Returns:
+    hits: HR score.
+    ndcgs: NDCG score.
+  """
+  predict_input_fn = tf.estimator.inputs.numpy_input_fn(
+      x={'user_input': user_input, 'item_input': item_input},
+      batch_size=100,
+      num_epochs=1,
+      shuffle=False
+  )
+  # Get predictions
+  predictions = est_model.predict(input_fn=predict_input_fn)
+  all_predicted_scores = [p['prediction'] for p in predictions]
+
+  # Calculate HR score
+  def _get_hr(ranklist, gt_item):
+    for item in ranklist:
+      if item == gt_item:
+        return 1
+    return 0
+
+  # Calculate NDCG score
+  def _get_ndcg(ranklist, gt_item):
+    for i in range(len(ranklist)):
```

Comment:
This can be updated to:

return 0 if gt_item not in ranklist else math.log(2) / math.log(ranklist.index(gt_item))

Reply:
The expression should be math.log(2) / math.log(ranklist.index(gt_item) + 2)?